src/cartesi-machine.lua

```diff
@@ -1708,70 +1595,59 @@ else
-                    cycles = machine:read_mcycle()
-                else
-                    assert(reason == cartesi.machine.HTIF_YIELD_REASON_RX_ACCEPTED, "invalid manual yield reason")
+
+            if reason == cartesi.machine.HTIF_YIELD_REASON_RX_REJECTED then
+                machine:rollback()
+                cycles = machine:read_mcycle()
+            else
+                assert(reason == cartesi.machine.HTIF_YIELD_REASON_RX_ACCEPTED, "invalid manual yield reason")
+
+                -- extract ouput_hash, skip first yield
+                if rollup_advance.next_input_index ~= 0 then
+                    assert(output_length == 32)
+                    save_rollup_state_output_hash(machine, config.rollup.tx_buffer, rollup_advance, output_length)
                 end
-                stderr("\nEpoch %d before input %d\n", rollup_advance.epoch_index, rollup_advance.next_input_index)
+            end
+            if rollup_advance and rollup_advance.next_input_index < rollup_advance.input_index_end then
                 if rollup_advance.hashes then print_root_hash(machine) end
                 machine:snapshot()
-                load_rollup_input_and_metadata(machine, config.rollup, rollup_advance)
+                local input_length = load_rollup_input(machine, config.rollup, rollup_advance)
                 if rollup_advance.hashes then print_root_hash(machine) end
                 machine:reset_iflags_Y()
-                machine:write_htif_fromhost_data(0) -- tell machine it is an rollup_advance state, but this is default
-                rollup_advance.voucher_index = 0
-                rollup_advance.notice_index = 0
-                rollup_advance.report_index = 0
+                machine:write_htif_fromhost_data((0 << 32) | input_length) -- tell machine it is an rollup_advance state, but this is default
+                rollup_advance.output_index = 0
                 rollup_advance.next_input_index = rollup_advance.next_input_index + 1
             else
-                -- there are outputs of a prevous advance state to save
-                if rollup_advance and rollup_advance.next_input_index > rollup_advance.input_index_begin then
-                    save_rollup_voucher_and_notice_hashes(machine, config.rollup, rollup_advance)
-                end
                 -- there is an inspect state query to feed
                 if rollup_inspect and rollup_inspect.query then
                     stderr("\nBefore query\n")
-                    load_rollup_query(machine, config.rollup, rollup_inspect)
+                    local input_length = load_rollup_query(machine, config.rollup, rollup_inspect)
                     machine:reset_iflags_Y()
-                    machine:write_htif_fromhost_data(1) -- tell machine it is an inspect state
-                    rollup_inspect.report_index = 0
+                    machine:write_htif_fromhost_data((1 << 32) | input_length) -- tell machine it is an inspect state
```

Comment:
If we add the constant to `htif-defines.h` and to the Lua bind,  we could use `cartesi.machine.HTIF_YIELD_REASON_INSPECT_STATE` here instead of `1`.